diff --git neofetch-clean/neofetch neofetch-workdir/neofetch
index 1e4b564..4224c55 100755
--- neofetch-clean/neofetch
+++ neofetch-workdir/neofetch
@@ -929,6 +929,7 @@ get_os() {
         AIX)      os=AIX ;;
         IRIX*)    os=IRIX ;;
         FreeMiNT) os=FreeMiNT ;;
+        Lyre)     os=Lyre ;;
 
         Linux|GNU*)
             os=Linux
@@ -954,6 +955,7 @@ get_distro() {
     [[ $distro ]] && return
 
     case $os in
+    	Lyre) distro=Lyre ;;
         Linux|BSD|MINIX)
             if [[ -f /bedrock/etc/bedrock-release && $PATH == */bedrock/cross/* ]]; then
                 case $distro_shorthand in
@@ -5288,6 +5290,27 @@ get_distro_ascii() {
     #
     # $ascii_distro is the same as $distro.
     case $(trim "$ascii_distro") in
+        "Lyre"*)
+            set_colors 4 1
+            read -rd '' ascii_data <<'EOF'
+${c1}
+########
+########
+########
+########
+########
+########
+########
+########
+########
+########
+########
+################
+################
+################
+################
+EOF
+	;;
         "AIX"*)
             set_colors 2 7
             read -rd '' ascii_data <<'EOF'
