diff --git mlibc-clean/abis/linux/in.h mlibc-workdir/abis/linux/in.h
index 3618abf..a4c4a35 100644
--- mlibc-clean/abis/linux/in.h
+++ mlibc-workdir/abis/linux/in.h
@@ -65,6 +65,10 @@ struct ipv6_mreq {
 #define INET_ADDRSTRLEN  16
 #define INET6_ADDRSTRLEN 46
 
+#define IPV6_MULTICAST_HOPS 18
+
+#define IP_MULTICAST_TTL 33
+
 #define IPPORT_RESERVED 1024
 
 #define IPPROTO_IP       0
diff --git mlibc-clean/options/glibc/generic/execinfo.cpp mlibc-workdir/options/glibc/generic/execinfo.cpp
index 3474615..de088e9 100644
--- mlibc-clean/options/glibc/generic/execinfo.cpp
+++ mlibc-workdir/options/glibc/generic/execinfo.cpp
@@ -1,17 +1,18 @@
 #include <execinfo.h>
 #include <bits/ensure.h>
+// #include <mlibc/debug.hpp>
 
-int backtrace(void **, int) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
-}
+// int backtrace(void **, int) {
+// 	mlibc::infoLogger() << "backtrace() is a stub" << frg::endlog;
+// 	return 0;
+// }
 
-char **backtrace_symbols(void *const *, int) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
-}
+// char **backtrace_symbols(void *const *, int) {
+// 	__ensure(!"Not implemented");
+// 	__builtin_unreachable();
+// }
 
-void backtrace_symbols_fd(void *const *, int, int) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
-}
+// void backtrace_symbols_fd(void *const *, int, int) {
+// 	__ensure(!"Not implemented");
+// 	__builtin_unreachable();
+// }
