diff --git mlibc-clean/abis/linux/in.h mlibc-workdir/abis/linux/in.h
index bb5b6ef..8d66fe8 100644
--- mlibc-clean/abis/linux/in.h
+++ mlibc-workdir/abis/linux/in.h
@@ -86,6 +86,10 @@ struct group_req {
 #define INET_ADDRSTRLEN  16
 #define INET6_ADDRSTRLEN 46
 
+#define IPV6_MULTICAST_HOPS 18
+
+#define IP_MULTICAST_TTL 33
+
 #define IPPORT_RESERVED 1024
 
 #define IPPROTO_IP       0
diff --git mlibc-clean/options/glibc/generic/execinfo.cpp mlibc-workdir/options/glibc/generic/execinfo.cpp
index 3474615..de088e9 100644
--- mlibc-clean/options/glibc/generic/execinfo.cpp
+++ mlibc-workdir/options/glibc/generic/execinfo.cpp
@@ -1,17 +1,18 @@
 #include <execinfo.h>
 #include <bits/ensure.h>
+// #include <mlibc/debug.hpp>
 
-int backtrace(void **, int) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
-}
+// int backtrace(void **, int) {
+// 	mlibc::infoLogger() << "backtrace() is a stub" << frg::endlog;
+// 	return 0;
+// }
 
-char **backtrace_symbols(void *const *, int) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
-}
+// char **backtrace_symbols(void *const *, int) {
+// 	__ensure(!"Not implemented");
+// 	__builtin_unreachable();
+// }
 
-void backtrace_symbols_fd(void *const *, int, int) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
-}
+// void backtrace_symbols_fd(void *const *, int, int) {
+// 	__ensure(!"Not implemented");
+// 	__builtin_unreachable();
+// }
diff --git mlibc-clean/options/posix/generic/termios-stubs.cpp mlibc-workdir/options/posix/generic/termios-stubs.cpp
index 3e621ad..4c4ff04 100644
--- mlibc-clean/options/posix/generic/termios-stubs.cpp
+++ mlibc-workdir/options/posix/generic/termios-stubs.cpp
@@ -49,9 +49,13 @@ int tcflow(int fd, int action) {
 	return 0;
 }
 
-int tcflush(int, int) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+int tcflush(int fd, int queue_selector) {
+	MLIBC_CHECK_OR_ENOSYS(mlibc::sys_tcflush, -1);
+	if(int e = mlibc::sys_tcflush(fd, queue_selector); e) {
+		errno = e;
+		return -1;
+	}
+	return 0;
 }
 
 int tcgetattr(int fd, struct termios *attr) {
diff --git mlibc-clean/options/posix/include/mlibc/posix-sysdeps.hpp mlibc-workdir/options/posix/include/mlibc/posix-sysdeps.hpp
index d56c0e4..eebfc7a 100644
--- mlibc-clean/options/posix/include/mlibc/posix-sysdeps.hpp
+++ mlibc-workdir/options/posix/include/mlibc/posix-sysdeps.hpp
@@ -147,6 +147,7 @@ int sys_vm_unmap(void *pointer, size_t size);
 [[gnu::weak]] int sys_tcgetattr(int fd, struct termios *attr);
 [[gnu::weak]] int sys_tcsetattr(int, int, const struct termios *attr);
 [[gnu::weak]] int sys_tcflow(int, int);
+[[gnu::weak]] int sys_tcflush(int, int);
 [[gnu::weak]] int sys_tcdrain(int);
 [[gnu::weak]] int sys_pipe(int *fds, int flags);
 [[gnu::weak]] int sys_socketpair(int domain, int type_and_flags, int proto, int *fds);
