diff --git sdl2-clean/CMakeLists.txt sdl2-workdir/CMakeLists.txt
index c5c9a77..8a9067f 100644
--- sdl2-clean/CMakeLists.txt
+++ sdl2-workdir/CMakeLists.txt
@@ -181,6 +181,8 @@ elseif(UNIX AND NOT APPLE)
     set(AIX TRUE)
   elseif(CMAKE_SYSTEM_NAME MATCHES "Minix.*")
     set(MINIX TRUE)
+  elseif(CMAKE_SYSTEM_NAME MATCHES "Lyre")
+    set(LYRE_TRUE)
   endif()
 elseif(APPLE)
   if(CMAKE_SYSTEM_NAME MATCHES ".*Darwin.*")
